swig/python/codegen/codegen.py

```diff
@@ -0,0 +1,1030 @@
+from __future__ import print_function, unicode_literals
+import six
+import sys
+import ast
+import os
+import tokenize
+import warnings
+from six import StringIO
+
+
+def interleave(inter, f, seq):
+    """Call f on each item in seq, calling inter() in between.
+    """
+    seq = iter(seq)
+    try:
+        f(next(seq))
+    except StopIteration:
+        pass
+    else:
+        for x in seq:
+            inter()
+            f(x)
+
+class CodeGenException(Exception):
+    """ Generic exception for errors raised in code generation """
+    pass
+
+class CodeGenerator:
+    """Methods in this class recursively traverse an AST and
+    output source code for the abstract syntax; original formatting
+    is disregarded. """
+
+    # represents built in functions
+    pythonbuiltins = ["abs", "float", "int"]
+    
+    # basic types
+    basic_arg_types = ['float', 'int']
+    
+    # supported math constansts    
+    mathconsts = {"pi": "M_PI",
+                  "e": "M_E",
+                  "inf": "INFINITY",
+                  "nan": "NAN",
+                  }
+    
+    # support for most numpy types except complex numbers and float>64bit
+    numpytypes = {"byte": "char",
+                  "ubyte": "unsigned char",
+                  "short": "short",
+                  "ushort": "unsigned short",
+                  "intc": "int",
+                  "uintc": "unsigned int",
+                  "uint": "unisgned int",
+                  "longlong": "long long",
+                  "ulonglong": "unsigned long long",
+                  "half": "half",       # cuda supported
+                  "single": "float",
+                  "double": "double",
+                  "longdouble": "long double",
+                  "bool_": "bool",
+                  "bool8": "bool",
+                  # sized aliases
+                  "int_": "long",
+                  "int8": "int8_t",
+                  "int16": "int16_t",
+                  "int32": "int32_t",
+                  "int64": "int64_t",
+                  "intp": "intptr_t",
+                  "uint_": "long",
+                  "uint8": "uint8_t",
+                  "uint16": "uint16_t",
+                  "uint32": "uint32_t",
+                  "uint64": "uint64_t",
+                  "uintp": "uintptr_t",
+                  "float_": "float",
+                  "float16": "half",
+                  "float32": "float",
+                  "float64": "double"
+                  }
+    
+    # getVariableType and setVariableType functions are added dynamically    
+    fgpu_funcs = [ "getID", "getStepCounter", "getThreadIndex" ]   
+    fgpu_attrs = ["ALIVE", "DEAD"]
+    fgpu_input_msg_funcs = ["getIndex"] 
+    fgpu_output_msg_funcs = []
+    fgpu_agent_out_msg_funcs = ["getID"]
+    fgpu_env_funcs = ["containsProperty"] # TODO: Get macro property
```

Comment:
TODO

Reply:
Macro properties are a feature that it maybe quite hard to support without considering much broader support for array types. I think it would be better to merge with this as a known limitation at this stage.